### ISSUE TITLE
BigQuery: make jobs awaitable

### ISSUE THREAD

**Issue by dkapitan** [googleapis/python-bigquery#18]
I know BigQuery jobs are asynchronous by default. However, I am struggling to make my datapipeline async end-to-end.

Looking at [this JS example](https://medium.com/skyshidigital/what-bigquery-is-and-isnt-d092fa173d53), I thought it would be the most Pythonic to make a BigQuery job awaitable. However, I can't get that to work in Python i.e. errors when `await client.query(query)`. Looking at the source code, I don't see which method [returns an awaitable object](https://www.python.org/dev/peps/pep-0492/).

 I have little experience in writing async Python code and found [this example that wraps jobs in a `async def coroutine`](https://stackoverflow.com/questions/53065340/python-how-to-big-query-asynchronous-tasks).
```
class BQApi(object):                                                                                                 
    def __init__(self):                                                                                              
        self.api = bigquery.Client.from_service_account_json(BQ_CONFIG["credentials"])                               

    async def exec_query(self, query, **kwargs) -> bigquery.table.RowIterator:                                       
        job = self.api.query(query, **kwargs)                                                                        
        task = asyncio.create_task(self.coroutine_job(job))                                                          
        return await task                                                                                            

    @staticmethod                                                                                                    
    async def coroutine_job(job):                                                                                    
        return job.result()   
```

The [`google.api_core.operation.Operation`](https://googleapis.github.io/google-cloud-python/latest/core/operation.html) shows how to use `add_done_callback` to asynchronously wait for long-running operations. I have tried that, but the following yields `AttributeError: 'QueryJob' object has no attribute '_condition'` :

```
from concurrent.futures import ThreadPoolExecutor, as_completed
query1 = 'SELECT 1'
query2 = 'SELECT 2'

def my_callback(future):
    result = future.result()

operations = [bq.query(query1), bq.query(query2)]
[operation.add_done_callback(my_callback) for operation in operations]
results2 = []
for future in as_completed(operations):
  results2.append(list(future.result()))
```

Given that jobs are already asynchronous, would it make sense to add a method that returns an awaitable?

Or am I missing something and is there an Pythonic way to use the BigQuery client with the async/await pattern?

**Comment by jsigee87** [comment]
Wanted to bump this thread, I am currently trying to write async BigQuery jobs myself and it would be great if the job were `awaitable`.

**Comment by vgrabovets** [comment]
Any news here? You've already dropped support for Python 2.

**Comment by yanhong-zhao-ef** [comment]
still waiting here... do you have an approximate timeline for this?

**Comment by tswast** [comment]
@yanhong-zhao-ef I'd love to have this feature, but it's going to take some design work to get right. In other Google Cloud libraries, asynchronous support is provided by a completely separate "AsyncClient". Since this BigQuery library is handwritten, I'd like to avoid having to maintain two copies of the same methods.

**Comment by tswast** [comment]
I talked with some folks on Firestore who've implemented this. They suggest a trio of classes: async_batch, base_batch, batch https://github.com/googleapis/python-firestore/tree/master/google/cloud/firestore_v1 in order to avoid too much duplicated effort.

**Comment by jarednash0** [comment]
> @dkapitan Thanks for the follow-up. I agree that it would be ideal if jobs could be `awaited`.
> 
> You could simplify your example code using a set, e.g.:
> 
> ```python
> from time import sleep
> 
> query1 = """
> SELECT
>   language.name,
>   average(language.bytes)
> FROM `bigquery-public-data.github_repos.languages` 
> , UNNEST(language) AS language
> GROUP BY language.name"""
> 
> query2 = 'SELECT 2'
> 
> queries = [query_1, query_2]
> 
> awaiting_jobs = set()
> 
> def callback(future):
>     awaiting_jobs.discard(future.job_id)
> 
> for query in queries:
>     job = bq.query(query)
>     awaiting_jobs.add(job.job_id)
>     job.add_done_callback(callback)
> 
> while awaiting_jobs:
>     print('waiting for jobs to finish ... sleeping for 1s')
>     sleep(1)
> 
> print('all jobs done, do your stuff')
> ```

I tried implementing this with copy_table, but I added "assert future.done()" as the first line in callback() and found that the assertion was failing. This must mean that the callback is getting executed before the job is actually done. Can you confirm this isn't the intended functionality?

**Comment by tswast** [comment]
> This must mean that the callback is getting executed before the job is actually done. Can you confirm this isn't the intended functionality?

@jarednash0  Definitely not intended. It'd be worth filing a separate issue for this so that we can address it, ideally with the code example that reproduces it.

**Comment by theophile-dh** [comment]
Any news on this issue? It would be a very useful feature

**Comment by eboddington** [comment]
Would definitely appreciate this as well

**Comment by adamserafini** [comment]
Any update from Google on `asyncio` compatibility for BigQuery lib?

**Comment by jasonwillschiu** [comment]
This is probably not the best, but right now a simple way is to chain the queries together in one mega(big)query haha.

My clunky example:

```
Q1 = f"""
  CREATE OR REPLACE TABLE `{project}.{dataset}.{table}`
  AS
  SELECT * EXCEPT({currency_pair}) 
  FROM `{project}.{dataset}.{table}`;
  """ 

Q2 = f"""
  CREATE OR REPLACE TABLE `{project}.{dataset}.{table}`
  AS
  (SELECT *
  FROM `{project}.{dataset}.{table}` tab1
  JOIN (
  SELECT {date_col}, {currency_pair}
  FROM `{project}.currency_conversion.{base_currency}_conversion_table` 
  ) currency
  ON tab1.date=currency.{date_col}
  );
  """

Chained_Query1 = Q1 + Q2
query_job = client.query(Chained_Query1)
```

**Comment by adamserafini** [comment]
AFAICT the above proposal isn't actually making the query awaitable though, is it? we want to run the query asynchronously, not synchronously.

**Comment by jasonwillschiu** [comment]
My bad, I thought my data was getting stuffed around because these bigquery processes were asynchronous already

**Comment by chalmerlowe** [comment]
Gonna close this out as "Will not work", due to conflicting priorities.

**Comment by kiraksi** [comment]
For awareness: This is an internship project and my internship ends in 2 weeks. I plan on implementing an AsyncClient for async_query_and_wait rpc, there has been other rpc's requested to be made asynchronously which I may not have time for. For future maintainers I will document my progress here, but please refer to the internal design doc on this feature which can be found at my internship documentation website go/kirnendra-internship

RPC methods to make async:
[] query_and_wait
[] query
[] get_job
[] get_query_results
[] get_table
[] list_partitions
[] get_partition

**Comment by chalmerlowe** [comment]
@adamserafini Thank you for your input.
We will take it into consideration.

**Comment by kiraksi** [comment]
New updates: there are some blocking issues in that in order to make a completely asynchronous RowIterator object for the query and query_and_wait methods, which are extensions of HTTPIterator in google-api-core library, I would need an AsyncHTTPIterator, I have optioned this issue here: https://github.com/googleapis/python-api-core/issues/627 and I may make this PR myself. However until then I will be focusing on the get_ methods that won't have this blocking issue.

**Comment by Tunneller** [comment]
Any news kiraksi? I guess that your internship has ended. Will the PR still go forward? Can you elaborate more on what we will see? You mention that your solution will not be completely asynchronous?

**Comment by tommyhutcheson** [comment]
Hey @tswast @chalmerlowe 
Is there an update for this issue and @kiraksi PR https://github.com/googleapis/python-bigquery/pull/1853 the python / BigQuery communities would love this feature to be available.

**Comment by chalmerlowe** [comment]
At this moment, this task is on hold. We will revisit it when we have manpower. I am not gonna close it so that it remains on the radar.

**Comment by pingyeh** [comment]
Before the async libraries are ready, it seems the done callback is the best way to run multiple queries in parallel unless the developer is willing to do multi-threading.

Can you please improve the documentation on [add_done_callback](https://cloud.google.com/python/docs/reference/bigquery/latest/google.cloud.bigquery.job.LoadJob#google_cloud_bigquery_job_LoadJob_add_done_callback)? It is unclear what the argument to the callback is, and what the callback can do with it. It's a `Future` class, but which `Future` class is it? Is it `asyncio.Future` or `google.api_core.future.base.Future` or something else?

**Comment by alexandernst** [comment]
I have just red the entire thread and I'm not sure if I'm reading comments written by paid devs working for one of the wealthiest organizations in the world (certainly one of the top 3 cloud providers) or by a couple of people working on their spare time on their least important side project.

I just refuse to believe that Google is capable of **neglecting** the official library for BigQuery in such a way. What a sorry state. 😞

**Comment by jacek-jablonski** [comment]
Any progress on this? Same goes for all other google-cloud-* libraries. Proper asyncio implementations are missing in the majority of the libraries. Lots of developers are struggling to integrate with Google Cloud from FastAPI or Starlette APIs.

**Comment by xin-hao-2025** [comment]
WE NEED THIS!!!

**Comment by bhuesemann** [comment]
This is 6 years old now. It will not come. There is no systematic support for async, and it seems to have no priority.

@kiraksi Have you heard anything about future plans for this?

**Comment by bhuesemann** [comment]
> There has been work to integrate BigQuery V2 APIs into https://github.com/googleapis/googleapis/tree/master/google/cloud/bigquery/v2
> 
> Theoretically, this could unlock the use of https://github.com/googleapis/gapic-generator-python which supports creating async clients.
> 
> CC [@shollyman](https://github.com/shollyman) and [@chalmerlowe](https://github.com/chalmerlowe) who have been pushing forward on many of these efforts. IMO even the low-level generated clients could be useful for such use cases.

@tswast Thanks for giving some feedback and reigniting hope that this topic might be resolved in the future.

**Comment by Tunneller** [comment]
Is this still pertinent now that Python has released the GIL ? I'm running 3.14t noGIL on Cloud Run with individual threads downloading from BQ and everything is in parallel. No async within the thread, but hey, I have 8 threads. And yes, much faster.
